packages/core/src/locations/types.ts

```diff
@@ -0,0 +1,55 @@
+/**
+ * A block id is a unique identifier for a block, it is a string.
+ */
+export type BlockId = string;
+
+/**
+ * A block identifier is a unique identifier for a block, it is either a string, or can be object with an id property (out of convenience).
+ */
+export type BlockIdentifier = { id: BlockId } | BlockId;
+
+/**
+ * A point is a path with an offset, it is used to identify a specific position within a block.
+ */
+export type Point = {
+  id: BlockId;
+  /**
+   * The number of characters from the start of the block.
+   * @note an offset of -1 is used to indicate that this point actually represents the whole block
+   */
+  offset: number;
+};
+
+/**
+ * A range is a pair of points, it is used to identify a range of blocks within a document.
+ */
+export type Range = {
+  anchor: Point;
```

Comment:
the term anchor / head comes from prosemirror selections, where the direction of a selection can matter.

For locations, I don't think direction matters, right? In that case, would start / from be better?

Reply:
I went with it like this to allow the direction to matter when it needs to, but be agnostic in most other cases. My only fear with making it be `from` & `to` is that I don't want to have to be finding the minimum between the two to make sure that the order is right all of the time.


For example:

```ts
editor.insertContentAt({ from: Math.min(resolveLocationToPm(point1), resolveLocationToPm(point2)), to: Math.max(resolveLocationToPm(point1), resolveLocationToPm(point2))

// v.s.

editor.insertContentAt({ from: point1, to: point2 })
```

Maybe, I'm thinking too much in terms of ProseMirror, and we can relax `from` being less than `to` as an invariant? That would still make sense to me, I guess, just a slight difference from ProseMirror behavior